src/infra/core/tests/tests/test_compact_service_impl.rs

```diff
@@ -0,0 +1,778 @@
+    let (_, last_old_block) = old_blocks.first().unwrap();
+    let (_, last_new_block) = new_blocks.first().unwrap();
+    assert!(CompactTestHarness::assert_add_data_block_events(
+        &last_old_block.event,
+        &last_new_block.event
+    ));
+
+    let data_str = indoc!(
+        "
+        date,city,population
+        2020-01-07,A,1000
+        2020-01-08,B,2000
+        2020-01-09,C,3000
+        "
+    );
+
+    harness
+        .ingest_data(data_str.to_string(), &root_dataset_alias.as_local_ref())
+        .await;
+
+    let old_blocks = harness
+        .get_dataset_blocks(&root_dataset_alias.as_local_ref())
+        .await;
+
+    assert_matches!(
+        harness
+            .compact_svc
+            .compact_dataset(
+                &dataset_handle,
+                MAX_SLICE_SIZE,
+                6,
+                Some(Arc::new(NullCompactionMultiListener {}))
+            )
+            .await,
+        Ok(CompactResult::Finished),
+    );
+
+    // Due to limit we will left last added slice as it is
+    data_helper
+        .assert_last_data_eq(
+            indoc!(
+                r#"
+                message arrow_schema {
+                  OPTIONAL INT64 offset;
+                  REQUIRED INT32 op;
+                  REQUIRED INT64 system_time (TIMESTAMP(MILLIS,true));
+                  OPTIONAL INT64 date (TIMESTAMP(MILLIS,true));
+                  OPTIONAL BYTE_ARRAY city (STRING);
+                  OPTIONAL INT64 population;
+                }
+                "#
+            ),
+            indoc!(
+                r#"
+                +--------+----+----------------------+----------------------+------+------------+
+                | offset | op | system_time          | date                 | city | population |
+                +--------+----+----------------------+----------------------+------+------------+
+                | 6      | 0  | 2050-01-01T12:00:00Z | 2020-01-07T00:00:00Z | A    | 1000       |
+                | 7      | 0  | 2050-01-01T12:00:00Z | 2020-01-08T00:00:00Z | B    | 2000       |
+                | 8      | 0  | 2050-01-01T12:00:00Z | 2020-01-09T00:00:00Z | C    | 3000       |
+                +--------+----+----------------------+----------------------+------+------------+
+                "#
+            ),
+        )
+        .await;
+
+    let new_blocks = harness
+        .get_dataset_blocks(&root_dataset_alias.as_local_ref())
+        .await;
+
+    // Shoud save original amount of blocks
+    // such as size is equal to max-slice-size
+    assert_eq!(old_blocks.len(), new_blocks.len());
+    let (_, last_old_block) = old_blocks.first().unwrap();
+    let (last_new_block_hash, last_new_block) = new_blocks.first().unwrap();
+    assert!(CompactTestHarness::assert_add_data_block_events(
+        &last_old_block.event,
+        &last_new_block.event
+    ));
+
+    harness
+        .commit_set_licence_block(&root_dataset_alias.as_local_ref(), last_new_block_hash)
```

Comment:
This test is getting very hard to read as it's checking for so many things. Can we split it into several:
- happy path - merge two blocks
- respect row/size limits
- non-data block in the middle

I would also like to see a test where blocks that are not at the head of the dataset get merged, e.g. we have a chain like:
`(data 1 small) <- (data 2 small) <- (data 3 large)`
and the compaction results in:
`(data 1,2 medium) <- (data 3 large`

Reply:
test added